src/target/llvm/codegen_params.cc

```diff
@@ -0,0 +1,151 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+/*!
+ * \file codegen_params.cc
+ */
+#ifdef TVM_LLVM_VERSION
+
+#include "codegen_params.h"
+
+#include <memory>
+#include <vector>
+
+namespace tvm {
+namespace codegen {
+
+namespace {
+class DLManagedTensorDeleter {
+ public:
+  void operator()(DLManagedTensor* ptr) { ptr->deleter(ptr); }
+};
+}  // namespace
+
+llvm::ConstantArray* NDArrayToLLVMArray(llvm::LLVMContext* ctx, ::tvm::runtime::NDArray arr) {
+  llvm::Type* element_type = nullptr;
+
```

Comment:
I wonder whether or not should we instead look into LLVM's target machine to lookup the endian parameter, and do byte swap ourselves.

This would allow us to generalizes to data types that goes beyond the types that can recognized by LLVM, as long as we know the target machine layout

See https://github.com/apache/incubator-tvm/blob/main/src/runtime/rpc/rpc_endpoint.cc#L429

Reply:
the thing is, because it will be used directly by instructions (I.e. load half-word), I prefer to let LLVM deal with alignment and whatever else is necessary to make that work. my understanding was we tried an experiment compiling large constants and the bulk of the time was spent in parsing, so this shouldn't actually load LLVM too much? also, we just do this a few times for the whole model

Reply:
How about we take the middle ground approach, e.g. using the same data type for the same bits. E.G. use uint32 for both float32, int32. This will allow us to effectively support new data types without having llvm to be aware of it

Reply:
won't we need to expand the switch-case statement to support new datatypes anyhow? for some reason, LLVM has a different constant impl for FP and int constants, so I guess i sort of like the approach taken now in the absence of any example new datatypes. i also don't think it's inflexible to supporting those new datatypes. note that LLVM does treat signed and unsigned constants as the same; just the codepath to instantiate them is slightly different.

Reply:
We still need to support cases like bfloat16, which is not natively supported by LLVM. It would still be great to support via uint in these cases

Reply:
added support for float16 and bfloat16 as discussed.